pgrx-pg-sys/src/submodules/elog.rs

```diff
@@ -96,195 +96,333 @@ impl From<i32> for PgLogLevel {
     }
 }
 
-/// Log to Postgres' `debug5` log level.
-///
-/// This macro accepts arguments like the [`println`] and [`format`] macros.
-/// See [`fmt`](std::fmt) for information about options.
-///
-/// The output these logs goes to the PostgreSQL log file at `DEBUG5` level, depending on how the
-/// [PostgreSQL settings](https://www.postgresql.org/docs/current/runtime-config-logging.html) are configured.
-#[macro_export]
-macro_rules! debug5 {
-    ($($arg:tt)*) => (
-        {
-            extern crate alloc;
-            $crate::ereport!($crate::elog::PgLogLevel::DEBUG5, $crate::errcodes::PgSqlErrorCode::ERRCODE_SUCCESSFUL_COMPLETION, alloc::format!($($arg)*).as_str());
+// A helper macro to define the various elog macros.
+macro_rules! define_elog {
+    (
+        $name:tt = {
+            loglevel = $loglevel:tt
+            errcode = $errcode:tt
+            docs = $doc_str:tt
+            extra_docs = $extra_docs:tt
+            extra_stmts = $extra_stmts:tt
+        }
+    ) => {
+        define_elog!{
+            @define_macro
+            name = $name
+            loglevel = $loglevel
+            errcode = $errcode
+            docs = $doc_str
+            extra_docs = $extra_docs
+            extra_stmts = $extra_stmts
+            dollar = $
+        }
+    };
+    (@get_first_doc_line name = $name:tt $doc_str:tt) => {
+        concat!("Log to Postgres' `", stringify!($name), "` log level. ", $doc_str)
+    };
+    (@get_extra_docs loglevel = $loglevel:tt false) => {""};
+    (@get_extra_docs loglevel = $loglevel:tt true) => {
+        concat!(
+            "The output these logs goes to the PostgreSQL log file at `", stringify!($loglevel), "` level, depending on how the\n",
+            "[PostgreSQL settings](https://www.postgresql.org/docs/current/runtime-config-logging.html) are configured.",
+            "\n"
+        )
+    };
+    (
+        @define_macro
+        name = $name:tt
+        loglevel = $loglevel:tt
+        errcode = $errcode:tt
+        docs = $doc_str:tt
+        extra_docs = $extra_docs:tt
+        extra_stmts = $extra_stmts:tt
+        dollar = $_:tt
+    ) => {
+        #[allow(non_snake_case)]
+        #[macro_export]
+        #[doc=concat!(
+            define_elog!(@get_first_doc_line name = $name $doc_str),
+            "\n",
+            "This macro accepts arguments like the [`println`] and [`format`] macros.\n",
+            "See [`fmt`](std::fmt) for information about options.\n",
```

Comment:
I worry about leaning on `fmt` syntax if we're going to use this `assignment = "literal", expr;` syntax which doesn't match the actual way fmt syntax is used?

Reply:
...Does this syntax currently support `message = "{message}", message = $expr;`?